### PR TITLE
(Typeahead) New configuration - Auto select item

### PR DESCRIPTION
**Pull request by kjartanvalur**
See issue #748

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@kjartanvalur thnx for the PR. In order to consider it for merging please add tests (you can use the existing ones as an example). Check our contrib guidelines for more info: https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md

**Comment by kjartanvalur** [comment]
I did add a test to typeahead-config.spec.ts :)

**Comment by kjartanvalur** [comment]
Updated tests, and fixed conflicts with other tests

File: src/typeahead/typeahead-window.ts [hunk 1]
```diff
@@ -73,7 +75,13 @@ export class NgbTypeaheadWindow {
 
   next() { this.activeIdx = (this.activeIdx + 1) % this.results.length; }
 
-  prev() { this.activeIdx = (this.activeIdx === 0 ? this.results.length - 1 : this.activeIdx - 1); }
+  prev() {
+    this.activeIdx =
+        ((this.autoSelectItem === true && this.activeIdx === 0) ||
+                 (this.autoSelectItem === false && this.activeIdx === -1) ?
+             this.results.length - 1 :
+             this.activeIdx - 1);
```
Comment (wesleycho) [inline src/typeahead/typeahead-window.ts]: There is way too much logic compacted in these lines - find a way to simplify readabiliity

**Comment by wesleycho** [comment]
This needs tests for all the conditional branching that it introduces in the parse function so the behavior is codified in tests.

File: src/typeahead/typeahead.ts [hunk 2]
```diff
@@ -108,6 +108,11 @@ export class NgbTypeahead implements OnInit,
   @Input() showHint: boolean;
 
   /**
+  * Auto select the first item in the result list
+  */
+  @Input() autoSelectItem: boolean = null;
```
Comment (maxokorokov) [inline src/typeahead/typeahead.ts]: I don't think you need initialization here, especially to `null`. I think it defaults to `true` in the config

File: src/typeahead/typeahead-window.ts [hunk 3]
```diff
@@ -36,7 +37,8 @@ export interface ResultTemplateContext {
   `
 })
 export class NgbTypeaheadWindow {
-  activeIdx = 0;
+  activeIdx: number = 0;
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead-window.ts]: Revert this change. We don't need type here is it is inferred.

File: src/typeahead/typeahead-window.ts [hunk 4]
```diff
@@ -36,7 +37,8 @@ export interface ResultTemplateContext {
...
+  autoSelectItem: boolean = true;
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead-window.ts]: We don't need type here is it is inferred. Remove `boolean`

File: src/typeahead/typeahead.spec.ts [hunk 5]
```diff
@@ -473,7 +473,7 @@ describe('ngb-typeahead', () => {
            const fixture = createTestComponent(`<input type="text" [(ngModel)]="model" 
                 [ngbTypeahead]="findAnywhere" 
                 [inputFormatter]="uppercaseFormatter" 
-                [showHint]="true"/>`);
+                [showHint]="true" [autoSelectItem]="true" />`);
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead.spec.ts]: We should _not_ modify existing tests. `autoSelectItem` should be default. Revert.

File: src/typeahead/typeahead.spec.ts [hunk 6]
```diff
@@ -496,7 +496,7 @@ describe('ngb-typeahead', () => {
 
       it('should restore hint when results window is dismissed', async(() => {
            const fixture = createTestComponent(
-               `<input type="text" [(ngModel)]="model" [ngbTypeahead]="findAnywhere" [showHint]="true"/>`);
+               `<input type="text" [(ngModel)]="model" [ngbTypeahead]="findAnywhere" [showHint]="true" [autoSelectItem]="true" />`);
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead.spec.ts]: We should _not_ modify existing tests. `autoSelectItem` should be default. Revert.

**Comment by pkozlowski-opensource** [comment]
As @wesleycho pointed out this needs much more tests, indeed. We should have at least:
* no item is selected with the new option switched on
* navigating with arrow down properly gets us to the next item
* navigating with arrow up / down properly wraps

File: src/typeahead/typeahead-window.ts (hunk 1 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead-window.ts]: done

File: src/typeahead/typeahead.ts (hunk 2 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead.ts]: done

File: src/typeahead/typeahead-window.ts (hunk 3 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead-window.ts]: done

File: src/typeahead/typeahead-window.ts (hunk 4 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead-window.ts]: done

File: src/typeahead/typeahead.spec.ts (hunk 5 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead.spec.ts]: done

File: src/typeahead/typeahead.spec.ts (hunk 6 above)
Comment (kjartanvalur) [inline src/typeahead/typeahead.spec.ts]: done

File: src/typeahead/typeahead.spec.ts [hunk 7]
```diff
@@ -473,7 +473,7 @@ describe('ngb-typeahead', () => {
            const fixture = createTestComponent(`<input type="text" [(ngModel)]="model" 
                 [ngbTypeahead]="findAnywhere" 
                 [inputFormatter]="uppercaseFormatter" 
-                [showHint]="true"/>`);
+                [showHint]="true" />`);
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead.spec.ts]: Why this change?

File: src/typeahead/typeahead.spec.ts [hunk 8]
```diff
@@ -525,21 +525,26 @@ describe('ngb-typeahead', () => {
...
         expect(typeahead.showHint).toBe(true);
+        expect(typeahead.autoSelect).toBe(false);
       });
     });
 
     describe('Custom config as provider', () => {
       beforeEach(() => {
         const config = new NgbTypeaheadConfig();
         config.showHint = true;
+        config.autoSelect = false;
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead.spec.ts]: I think that the more interesting test would exercise non-default path

File: src/typeahead/typeahead.ts [hunk 9]
```diff
@@ -147,6 +153,10 @@ export class NgbTypeahead implements OnInit,
             this._closePopup();
           } else {
             this._openPopup();
+            if (this.autoSelect !== null && this.autoSelect !== undefined) {
+              this._windowRef.instance.activeIdx = this.autoSelect ? 0 : -1;
+              this._windowRef.instance.autoSelect = this.autoSelect;
```
Comment (pkozlowski-opensource) [inline src/typeahead/typeahead.ts]: Based on my previous comment I don't think we need `this._windowRef.instance.autoSelect = this.autoSelect`

**Comment by pkozlowski-opensource** [comment]
@kjartanvalur I did another pass on this PR and unfortunately it is not in the state where we could consider it for merging :-( 

Besides comments in the code the PR doesn't merge cleanly (needs rebasing) and we are still tests for several use-cases I've brought up [previously](https://github.com/ng-bootstrap/ng-bootstrap/pull/767#issuecomment-249209093)

Given all the above and the fact that this PR haven't seen any activity in almost 2 weeks I'm going to close it for now. Feel free to open a new PR when you've got time to work on this. Thnx!
